Implement unitless length quirk

nox opened the pull request:
The Gecko side doesn't propagate its quirks mode yet.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16609)
<!-- Reviewable:end -->

nox committed:
Make Stylist::quirks_mode hold a QuirksMode

nox committed:
Propagate quirks mode all the way to ParserContext

The quirks mode is still not properly propagated in geckolib.

nox committed:
Implement the unitless length quirk for background-position

nox committed:
Implement the unitless length quirk for border-spacing

nox committed:
Enable /quirks-mode/ tests

nox committed:
Implement the unitless length quirk for border-*-width

nox committed:
Implement the unitless length quirk for border-width

nox committed:
Implement the unitless length quirk for position properties

nox committed:
Implement the unitless length quirk for clip

nox committed:
Implement the unitless length quirk for font-size

nox committed:
Implement the unitless length quirk for physical size properties

nox committed:
Implement the unitless length quirk for letter-spacing

nox committed:
Implement the unitless length quirk for margin-*

nox committed:
Implement the unitless length quirk for margin

nox committed:
Implement the unitless length quirk for physical size extremums

nox committed:
Implement the unitless length quirk for padding-*

nox committed:
Implement the unitless length quirk for padding

nox committed:
Implement the unitless length quirk for text-indent

nox committed:
Implement the unitless length quirk for word-spacing

nox committed:
Run a test per property and value in unitless-length.html

nox committed:
Implement the unitless length quirk for vertical-align

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/stylesheets.rs, components/style/values/specified/mod.rs, components/style/values/computed/mod.rs, components/style/properties/longhand/position.mako.rs, components/style/values/specified/length.rs and 28 more
 * @jgraham: tests/wpt/include.ini
 * @kichjang: components/script/dom/htmlmetaelement.rs, components/script/dom/window.rs, components/script/dom/css.rs, components/script/dom/document.rs, components/script/dom/htmllinkelement.rs and 9 more
 * @fitzgen: components/script/dom/htmlmetaelement.rs, components/script/dom/window.rs, components/script/dom/css.rs, components/script/dom/document.rs, components/script/dom/htmllinkelement.rs and 9 more
 * @emilio: ports/geckolib/glue.rs, components/style/stylesheets.rs, components/style/values/specified/mod.rs, components/style/values/computed/mod.rs, components/style/properties/longhand/position.mako.rs and 29 more

Manishearth commented on components/style/values/specified/length.rs:
```diff
@@ -1154,20 +1154,20 @@ impl LengthOrPercentage {
-                }
-                Ok(LengthOrPercentage::Length(NoCalcLength::Absolute(AbsoluteLength::Px(value.value))))
-            }
+            Token::Number(ref value) if value.value == 0. ||
```
nit: parehtneses over precedence

Manishearth commented:
r=me with those parens set.

nox commented:
@bors-servo r=Manishearth p=17

bors-servo commented:
:pushpin: Commit 25d2fbb has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 25d2fbb9ab082c13f2fc9407695a59ea1d76164e -->

bors-servo commented:
:hourglass: Testing commit 25d2fbb9ab082c13f2fc9407695a59ea1d76164e with merge 83884ece79ed3da5383487c914f03a9035e19f2c...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2073)

nox commented:
@bors-servo r=Manishearth

bors-servo commented:
:pushpin: Commit 65a0d89 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 65a0d8943e810c6bfce752132d268501a5cb44b1 -->

bors-servo commented:
:hourglass: Testing commit 65a0d8943e810c6bfce752132d268501a5cb44b1 with merge 96d896a1ab0acd033375cb6e8850c88512f00be3...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2074)

nox commented:
@bors-servo r=Manishearth

bors-servo commented:
:pushpin: Commit c170b96 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth c170b963dfa8e2fd7166b3d23020b15889891698 -->

bors-servo commented:
:hourglass: Testing commit c170b963dfa8e2fd7166b3d23020b15889891698 with merge 6c5df720409a097e4dfbf21366d76bf590932033...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3655)

Wafflespeanut commented:
Haste Haste Haste... Test expectations need to be updated.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16455) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
Well this is good news, turns out all the tests pass on Linux, and the failing tests on macOS are unrelated.

@bors-servo r=Manishearth

bors-servo commented:
:pushpin: Commit ce8a90e has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth ce8a90ec665f840e8f6134d6323e57112c701068 -->

bors-servo commented:
:hourglass: Testing commit ce8a90ec665f840e8f6134d6323e57112c701068 with merge c2a5742c0a030f366a37bd6b98f46b30ad5872e3...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/6899)

nox commented:
@bors-servo r=Manishearth

Tidy issue when rebasing.

bors-servo commented:
:pushpin: Commit 3364e27 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 3364e278dcf21bdc9ba3056d1056c060c991957b -->

bors-servo commented:
:hourglass: Testing commit 3364e278dcf21bdc9ba3056d1056c060c991957b with merge ed2850f19a24e589d930c574cc7fe1772953f67a...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/6900)

nox commented:
@bors-servo r=Manishearth

Build machine issues.

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.

bors-servo commented:
:pushpin: Commit 3364e27 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 3364e278dcf21bdc9ba3056d1056c060c991957b -->

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/6552), [arm32](http://build.servo.org/builders/arm32/builds/6762), [arm64](http://build.servo.org/builders/arm64/builds/6797), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6901), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3590), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2086) are reusable. Rebuilding only [linux-dev](http://build.servo.org/builders/linux-dev/builds/6900), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3638), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3665), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6939), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3659)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3665)

nox commented:
I'm not sure whether `/html/semantics/interactive-elements/the-dialog-element/centering.html` is intermittent or not. I hope it is.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 6dc589780d228cf42c2d82813b30fab9c9de7714 with merge 348a2d103d6549ccaa240722829d097df289c515...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3669)

nox commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 6dc589780d228cf42c2d82813b30fab9c9de7714 with merge 9b5ae4209c55632e866a1f9877f8f5e44b932a91...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3670)

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 16363629ee8a7ab8a4f8fcd21037649ce3ea815a with merge 3544c599666d3a336660193dd7eff6ee0684e7f5...

nox commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 16363629ee8a7ab8a4f8fcd21037649ce3ea815a with merge b4f52da593d38355342b03ebdef468e8e52ad84b...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3657)

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b5da2234b7c008813752c752892b4f0d6d937b4b with merge 0c4cef4d7b8aaa957e96213aee1e433ea9fcbc2d...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3689)

nox committed:
Copy the damn unitless-length.html test in /_mozilla/

It's the exact same thing as the one in WPT, but without the intermittent
'top' and 'bottom' failures which are unrelated to the damn quirk.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 1091af6fee5bd69165417bbcf19d66e2ea7c3d84 with merge 80bde98999a36a4789175608ca3b212971036db7...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/6583), [arm32](http://build.servo.org/builders/arm32/builds/6793), [arm64](http://build.servo.org/builders/arm64/builds/6826), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6931), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3668), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3695), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6932), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6967), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3621), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3689), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2118)
State: approved= try=True

nox commented:
@bors-servo r=Manishearth,emilio

bors-servo commented:
:pushpin: Commit 1091af6 has been approved by `Manishearth,emilio`

<!-- @bors-servo r=Manishearth,emilio 1091af6fee5bd69165417bbcf19d66e2ea7c3d84 -->

bors-servo commented:
:hourglass: Testing commit 1091af6fee5bd69165417bbcf19d66e2ea7c3d84 with merge 718c54e05721614627762dc042d8e10dc53a72b7...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3700)

nox commented:
@bors-servo retry #16627

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/6588), [arm32](http://build.servo.org/builders/arm32/builds/6798), [arm64](http://build.servo.org/builders/arm64/builds/6830), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6937), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3674), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6937), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6970), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3692), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2124) are reusable. Rebuilding only [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3700), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3626)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3703)

nox commented:
@bors-servo retry #16627

bors-servo commented:
:hourglass: Testing commit 1091af6fee5bd69165417bbcf19d66e2ea7c3d84 with merge f2bc8b004ab68419b5d0595cd746bf3362038975...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3679)

nox commented:
Build is broken?

@bors-servo retry

bors-servo commented:
:hourglass: Testing commit 1091af6fee5bd69165417bbcf19d66e2ea7c3d84 with merge f2c76c48038859764abe180bc37d439da07d2c3f...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3681)

nox commented:
@bors-servo retry #16631

bors-servo commented:
:hourglass: Testing commit 1091af6fee5bd69165417bbcf19d66e2ea7c3d84 with merge 922408434f73095c204a4a2242b0c929860c8045...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3683)

nox commented:
Build is broken?

@bors-servo retry

chenpighead commented:
I believe you're hitting #16632... which I also hit it for hours.... :-(

bors-servo commented:
:hourglass: Testing commit 1091af6fee5bd69165417bbcf19d66e2ea7c3d84 with merge 1d4cb4b106faf51d6ba56bb84bd98bc395ff38f5...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/6947)

bors-servo commented:
:hourglass: Testing commit 1091af6fee5bd69165417bbcf19d66e2ea7c3d84 with merge bdf74efef50edd16f7add0ad63c0f9d94fd85c76...

bors-servo commented:
:broken_heart: Test failed - [arm32](http://build.servo.org/builders/arm32/builds/6815)

jdm commented:
@bors-servo: retry

bors-servo commented:
:hourglass: Testing commit 1091af6fee5bd69165417bbcf19d66e2ea7c3d84 with merge d8bcc0db1aad26e007b7e2bdeda3cea4953c0db0...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/6614), [arm32](http://build.servo.org/builders/arm32/builds/6824), [arm64](http://build.servo.org/builders/arm64/builds/6855), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6964), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3701), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3727), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6964), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6997), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3652), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3720), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2150)
Approved by: Manishearth,emilio
Pushing d8bcc0db1aad26e007b7e2bdeda3cea4953c0db0 to master...
